### DISCUSSION TITLE
Support any OpenID Policy variable

### DISCUSSION THREAD

**Discussion by ribeaud** [minio/minio#17164] [Q&A]
**Is your feature request related to a problem? Please describe.**
I think it would be nice to be able to specify _any_ policy variable name, and not to be limited to the list specified [here](https://min.io/docs/minio/linux/administration/identity-access-management/policy-based-access-control.html#policy-variables). I do not understand this limitation by the way. Is it security relevant?

**Describe the solution you'd like**
In my concrete example, I am controlling the OIDC server. I can specify in the client configuration which claims should be sent to **MinIO** as part of the token.

In an ideal world, following policy could be used/defined:
```json
{
"Version": "2012-10-17",
"Statement": [
      {
         "Action": ["s3:ListBucket"],
         "Effect": "Allow",
         "Resource": ["arn:aws:s3:::${jwt:project}"],
         "Condition": {"StringLike": {"s3:prefix": ["${jwt:data_provider}/*"]}}
      },
      {
         "Action": [
         "s3:GetObject",
         "s3:PutObject"
         ],
         "Effect": "Allow",
         "Resource": ["arn:aws:s3:::${jws:project}/${jws:data_provider}/*"]
      }
   ]
}
```

**Describe alternatives you've considered**
The alternative is to strictly stick to the set of **jwt** namespace variables. So you will map _name_ to _data_provider_ but this feels like a hack in my case and does not fully reflect the reality.

**Comment by harshavardhana** [comment]
Policy conditions must be standard conditions, if you make it free-form anyone can mimic and spoof your conditions. 

We can alternatively take inputs asking for additional keys that are set by the user and we can use those to additionally replace and filter. 

This makes our policy JSON even more dynamic than what its already is, we would like to avoid such complexities in critical pieces in AuthZ. 

Current values relying on standard JWT spec are enough and cover most of the areas. Feel free to work on this and send a PR if you must we will take a look and review.

But this is currently out of scope.

**Comment by ribeaud** [comment]
Thanks for your answer which sounds reasonable to me. So, maybe, I am handling my use case the wrong way? 🤔 

Our policies are not based on _any_ of the predefined list, neither user's `preferred_username`, nor user's `website` or `phone_number` (what strange ones by the way...).

The closest one I could envisage is `jwt:groups` but I do not know how to use it. Is that an array? Or a simple string? I could not find any example in the documentation.

**Comment by ravindk89** [comment]
@ribeaud to clarify, what is your goal state here?  Are you trying to isolate users to specific prefixes or something similar?

w.r.t. `jwt:groups` that just depends on what you define upstream, IIRC. So with Keycloak if I map that claim to an array somehow, it would be an array. How exactly that would work depends on how you use it as a condition key.

https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_multi-value-conditions.html#reference_policies_multi-value-conditions-examples

AWS has some guidance here, but I've not tested something like that w/ MinIO.

**Comment by alexisdondon** [comment]
A policy like this could do the job no?

```
{
"Version": "2012-10-17",
"Statement": [
      {
         "Action": ["s3:ListBucket"],
         "Effect": "Allow",
         "Resource": ["arn:aws:s3:::projectName"],
          "Condition": {
        "ForAnyValue:StringEquals": {
          "jwt:groups": ["groupName"]
        }
      }
      },
      {
         "Action": [
         "s3:GetObject",
         "s3:PutObject"
         ],
         "Effect": "Allow",
         "Resource": ["arn:aws:s3:::projectName/someSubPath/*"],
   "Condition": {
        "ForAnyValue:StringEquals": {
          "jwt:groups": ["groupName"]
        }
      }
   ]
}

```
From What i understand the policy can't be generic:
- jwt:groups is based on an array. 
- projectName and read/modify a specific subPath, both need to be hardcoded in the policy.it is allowed if a specific groupName is present in array groups.

The drawback is to manage one policy by project and i don't know if minio have limit on number of policies ( i know that listing policy as admin is very slow with the number of policis but maybe there is no limit).

**Comment by jimrobinson** [comment]
I was also surprised that the set was constrained, since one of the interesting capabilities of using an OIDC is a looser coupling of authorization logic, where claims returned by an IdP may vary over time and each relaying party may alter which claims they recognize over time.

Stanford University offers an OIDC provider that includes a claim named “eduperson_entitlement” which corresponds to a field defined in the SAML based specification

https://wiki.refeds.org/display/STAN/eduPerson+2020-01#eduPerson202001-eduPersonEntitlement

Unfortunately this hasn’t yet made its way into

https://www.iana.org/assignments/jwt/jwt.xhtml

and I’m not sure Stanford or others will ever take steps to register it.

The discussion here strongly suggests to me that MinIO isn’t keen on expanding the list outside of the IANA registered list, and I see you discuss several concerns you have with expanding policy variable capabilities.

Since we are hoping to use MinIO and authenticate via Stanford we are exploring ways we can work around the current limitations.  

Changing the code was one obvious solution (anything from a simple edit of https://github.com/minio/pkg/blob/main/policy/condition/keyname.go to a more complicated change to register custom claims via configuration), but it wouldn’t be at the top of our list unless we could do it in a way that had a reasonable chance of being folded back into MinIO.

We do have some alternative solutions in mind, such as using keycloak or similar software to act as a proxy that could rewrite claims such as “eduperson_entitlement” to “groups”, but we thought we should ask for input from you before we proceed toward any specific solution.

**Reply by ravindk89** [reply]
cc @donatello in case you have guidance here.

In a previous comment:

> Current values relying on standard JWT spec are enough and cover most of the areas. Feel free to work on this and send a PR if you must we will take a look and review.

I don't think we'd reject the PR outright, but the concern originally was constraining complexity of policy logic as much as possible (IAM policies are already an entire _thing_ that people get caught on)

If you have a business case here, I would actually recommend reaching out to hello@min.io to discuss the requirements around the Stanford OIDC provider.
